server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -85,20 +86,13 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             .expandToAllNodes()) {
             return YES_AUTO_EXPAND_ALL;
         }
-        if (node.node() != null) {
-            assert Strings.hasLength(node.node().getHostAddress()) : node;
-            for (RoutingNode checkNode : allocation.routingNodes()) {
-                if (checkNode.node() == null) {
-                    continue;
-                }
-                // check if its on the same host as the one we want to allocate to
-                assert Strings.hasLength(checkNode.node().getHostAddress()) : checkNode;
-                if (checkNode.node().getHostAddress().equals(node.node().getHostAddress())) {
-                    for (ShardRouting assignedShard : assignedShards) {
-                        if (checkNode.nodeId().equals(assignedShard.currentNodeId())) {
-                            return allocation.debugDecision() ? debugNoAlreadyAllocatedToHost(node, allocation) : Decision.NO;
-                        }
-                    }
+        Hosts hosts = allocation.getHosts();
+        Host host = hosts.getHost(node.nodeId());
+        // return yes if can not find the host info of candidate node(no host name & address)
```

Comment:
I think we are guaranteed to have a host now since there always is a host address? Can we assert that instead?